lib/pkcs11/pkcs11_key.c

```diff
@@ -373,6 +373,61 @@ static CK_RV pkcs11_key_auth_required(CK_VOID_PTR pObject, CK_ATTRIBUTE_PTR pAtt
     return rv;
 }
 
+static CK_RV pkcs11_key_get_id(CK_VOID_PTR pObject, CK_ATTRIBUTE_PTR pAttribute)
+{
+    pkcs11_object_ptr obj_ptr = (pkcs11_object_ptr)pObject;
+    CK_RV rv = CKR_ARGUMENTS_BAD;
+
+    if (obj_ptr)
+    {
+#if PKCS11_AUTO_ID_ENABLE
+        if (pAttribute->pValue)
+        {
+            CK_BBOOL is_private;
+
+            if (CKR_OK == (rv = pkcs11_object_is_private(obj_ptr, &is_private)))
+            {
+                ATCA_STATUS status;
+                uint8_t buffer[1 + ATCA_ECCP256_PUBKEY_SIZE] = {0x04};
+
+                if (is_private)
+                {
+                    status = atcab_get_pubkey(obj_ptr->slot, &buffer[1]);
+                    PKCS11_DEBUG("atcab_get_pubkey: %x\r\n", status);
+                }
+                else
+                {
+                    status = atcab_read_pubkey(obj_ptr->slot, &buffer[1]);
+                    PKCS11_DEBUG("atcab_read_pubkey: %x\r\n", status);
+                }
+
+                if (ATCA_SUCCESS == status)
+                {
+                    status = atcac_sw_sha1(buffer, sizeof(buffer), buffer);
+                }
+
+                if (ATCA_SUCCESS == status)
+                {
+                    rv = pkcs11_attrib_fill(pAttribute, buffer, ATCA_SHA1_DIGEST_SIZE);
+                }
+                else
+                {
+                    rv = pkcs11_util_convert_rv(status);
+                }
+            }
+        }
+        else
+        {
+            rv = pkcs11_attrib_fill(pAttribute, NULL, ATCA_SHA1_DIGEST_SIZE);
```

Comment:
If pValue is null, this needs to fill in expected length, I'll send an update PR.

Reply:
pkcs11_attrib_fill is a utility function that has all of the require logic implemented: https://github.com/MicrochipTech/cryptoauthlib/blob/main/lib/pkcs11/pkcs11_attrib.c#L52-L91 so there is no need to check pAttribute->pValue or to manually set the length under most circumstances.